### PR TITLE
Support for 3rd party metrics

### PR DESCRIPTION
**Pull request by svanscho**
Added support for:
- metrics interval changes
- 3rd party metrics (source != openmotics)
- multiple metric types

### CODE REVIEW

**Comment by svanscho** [comment]
@khenderick let's discuss

File: src/vpn_service.py [hunk 1]
```diff
@@ -17,13 +17,14 @@
 if required. On each check the vpn_service sends some status information about the outputs and
 thermostats to the cloud, to keep the status information in the cloud in sync.
 """
-
+from bus.om_bus_client import MessageClient
+from bus.om_bus_events import Events
```
Comment (khenderick) [inline src/vpn_service.py]: Always do imports after `...import_eggs()`

File: src/led_service.py [hunk 2]
```diff
@@ -263,15 +264,20 @@ def main():
         led_controller.start()
         led_controller.set_led(Hardware.Led.POWER, True)
 
-        DBusService('led_service',
-                    event_receiver=lambda *args, **kwargs: led_controller.event_receiver(*args, **kwargs),
-                    get_state=led_controller.get_state)
+        message_service = MessageService('led_service')
+        message_service.add_event_handler(lambda *args, **kwargs: led_controller.event_receiver(*args, **kwargs))
```
[the remark this answers is not in the document]
Comment (svanscho) [reply src/led_service.py]: Yes

File: src/openmotics_service.py [hunk 3]
```diff
@@ -18,14 +18,17 @@
 
 import logging
 import time
-import threading
 
+from bus.om_bus_service import MessageService
+from bus.om_bus_client import MessageClient
+from bus.om_bus_events import Events
```
[the remark this answers is not in the document]
Comment (svanscho) [reply src/openmotics_service.py]: Ok

File: src/bus/om_bus_client.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        if msg['type'] == 'state':
+            self.latest_state_received = msg
+        if msg['type'] == 'event':
+            for callback in self.callbacks:
+                callback(msg['data']['event_type'], msg['data']['payload'])
+
+    def _message_receiver(self):
+        self.client = Client(self.address, authkey=self.authkey)
+        while True:
+            try:
```
[the remark this answers is not in the document]
Comment (svanscho) [reply src/bus/om_bus_client.py]: Agreed

File: src/bus/om_bus_client.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                msg = self.client.recv_bytes()
+                self._process_message(msg)
+            except EOFError:
+                self.client.close()
+                self.client = Client(self.address, authkey=self.authkey)
```
[the remark this answers is not in the document]
Comment (svanscho) [reply src/bus/om_bus_client.py]: Abstracted the timeout in a connect() method with self._connected check flag

File: src/bus/om_bus_service.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            target_client_name = msg['data']['client']
+            self._unicast(target_client_name, msg)
+        else:
+            logger.warning('Unknown message type: {0}'.format(msg_type))
+
+    def _receiver(self, conn):
+        try:
+            while True:
+                payload = conn.recv_bytes()
+                msg = json.loads(payload)
```
Comment (khenderick) [inline src/bus/om_bus_service.py]: Make sure to catch e.g. json parse exceptions

File: src/gateway/metrics.py [hunk 7]
```diff
@@ -258,13 +263,34 @@ def receiver(self, metric):
...
-        metric_types = self._config_controller.get_setting('cloud_metrics_types')
-        if metric_type not in metric_types:
-            return
 
         if self._config_controller.get_setting('cloud_enabled', True) is False:
             return
-        if self._config_controller.get_setting('cloud_metrics_enabled|{0}'.format(metric_type), False) is False:
+
+        if metric_source == 'OpenMotics':
+            # OM717: only energy and counter types can be opt-out atm
```
Comment (khenderick) [inline src/gateway/metrics.py]: Remove those, see other review remark

File: src/led_service.py [hunk 8]
```diff
@@ -14,29 +14,41 @@
...
 from ConfigParser import ConfigParser
 
-from bus.dbus_service import DBusService
-from bus.dbus_events import DBusEvents
+from signal import signal, SIGTERM
+from bus.om_bus_events import OMBusEvents
+from bus.om_bus_client import MessageClient
 from platform_utils import Hardware
 import constants
+import logging
```
Comment (khenderick) [inline src/led_service.py]: Import order

File: src/vpn_service.py [hunk 9]
```diff
@@ -401,82 +423,86 @@ def _unload_queue(queue):
...
         self._vpn_open = is_running and self._vpn_controller.vpn_connected
-        self._dbus_service.send_event(DBusEvents.VPN_OPEN, self._vpn_open)
+        self._message_client.send_event(OMBusEvents.VPN_OPEN, self._vpn_open)
 
     def start(self):
-        mainloop = gobject.MainLoop()
-        gobject.timeout_add(250, self._check_vpn)
-        mainloop.run()
+        self._check_vpn()
+        # t_check_vpn = Thread(target=self._check_vpn)
```
Comment (khenderick) [inline src/vpn_service.py]: Remove commented code

**Comment by svanscho** [comment]
Should be good to go now.

File: src/gateway/metrics.py [hunk 10]
```diff
@@ -346,8 +379,6 @@ def receiver(self, metric):
                     else:
                         self._cloud_retry_interval = 60 * 60
                         new_interval = 2 * 60 * 60
```
Comment (khenderick) [inline src/gateway/metrics.py]: PyCharm notifies about an unused variable. This variable is used to throttle down the metric collector if metrics can't be send to the cloud anyway. We still need that.

File: src/gateway/metrics.py [hunk 11]
```diff
@@ -482,11 +513,17 @@ def _distribute_openmotics(self):
...
                     self.outbound_rates[rate_key] += 1
                     self.outbound_rates['total'] += 1
             except IndexError:
                 time.sleep(0.1)
+
+    def event_receiver(self, event, payload):
+        if event == OMBusEvents.METRICS_INTERVAL_CHANGE:
+            for metric_type, interval in payload.iteritems():
+                self.set_cloud_interval(metric_type, interval)
+
```
Comment (khenderick) [inline src/gateway/metrics.py]: This line isn't needed

File: src/bus/om_bus_client.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
+import random
```
Comment (khenderick) [inline src/bus/om_bus_client.py]: Unused import

File: src/bus/om_bus_client.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+from multiprocessing.connection import Client
+from threading import Thread, Lock
+import logging
```
Comment (khenderick) [inline src/bus/om_bus_client.py]: Order of imports:
1. `import external_foo`
2. `from external.something import bar`
3. `from own.library import abc`

If there's something like trying an import and importing another module if needed, it should be placed on the location the plain import would have been.

File: src/bus/om_bus_service.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+import time
+
+try:
+    import json
+except ImportError:
+    import simplejson as json
+
+logger = logging.getLogger('openmotics')
+
+class MessageService():
```
Comment (khenderick) [inline src/bus/om_bus_service.py]: Two empty lines before a new class definition

File: src/gateway/gateway_api.py [hunk 15]
```diff
@@ -31,6 +31,9 @@
 import subprocess
 import tempfile
 import ConfigParser
+
+from bus.om_bus_events import OMBusEvents
```
Comment (khenderick) [inline src/gateway/gateway_api.py]: Move this to where the old events were imported

File: src/gateway/webservice.py [hunk 16]
```diff
@@ -35,7 +35,6 @@
 from ws4py.server.cherrypyserver import WebSocketPlugin, WebSocketTool
 from ws4py.websocket import WebSocket
 
-from bus.dbus_events import DBusEvents
```
Comment (khenderick) [inline src/gateway/webservice.py]: This wasn't replaced by an import of the new Events, breaking the `indicate` API at `webservice.py:2511`

File: src/vpn_service.py [hunk 17]
```diff
@@ -32,12 +32,13 @@
 import constants
 import threading
 
+from threading import Thread
 from collections import deque
 from ConfigParser import ConfigParser
 from datetime import datetime
```
Comment (khenderick) [inline src/vpn_service.py]: Unused import

File: src/vpn_service.py [hunk 18]
```diff
@@ -21,9 +21,9 @@
 from platform_utils import System
 System.import_eggs()
 
+import logging
 import os
 import glob
-import gobject
 import sys
```
Comment (khenderick) [inline src/vpn_service.py]: Unused import

File: src/vpn_service.py [hunk 19]
```diff
@@ -47,13 +48,19 @@
...
 
-class LOGGER(object):
-    @staticmethod
-    def log(line):
-        sys.stdout.write('{0}: {1}\n'.format(datetime.now(), line))
-        sys.stdout.flush()
 
+def setup_logger():
+    """ Setup the OpenMotics logger. """
+    logger = logging.getLogger("openmotics")
```
Comment (khenderick) [inline src/vpn_service.py]: You can just use `logger` as it exists in the module. It removes the PyCharm warning and you're not shadowing other variables.

**Comment by svanscho** [comment]
@khenderick

**Pull request merged**